source/f5_virtualserver.go

```diff
@@ -0,0 +1,412 @@
+
+	vs.virtualServerInformer.Informer().AddEventHandler(eventHandlerFunc(handler))
+}
+
+// endpointsFromVirtualServers extracts the endpoints from a slice of VirtualServers
+func (vs *f5VirtualServerSource) endpointsFromVirtualServers(virtualServers []*VirtualServer) ([]*endpoint.Endpoint, error) {
+	var endpoints []*endpoint.Endpoint
+
+	for _, virtualServer := range virtualServers {
+		ttl, err := getTTLFromAnnotations(virtualServer.Annotations)
+		if err != nil {
+			return nil, err
+		}
+
+		if virtualServer.Spec.VirtualServerAddress != "" {
+			ep := &endpoint.Endpoint{
+				Targets: endpoint.Targets{
+					virtualServer.Spec.VirtualServerAddress,
+				},
+				RecordType: "A",
+				DNSName:    virtualServer.Spec.Host,
+				Labels:     endpoint.NewLabels(),
+				RecordTTL:  ttl,
+			}
+
+			vs.setResourceLabel(virtualServer, ep)
+			endpoints = append(endpoints, ep)
+			continue
+		}
+
+		if virtualServer.Status.VSAddress != "" {
+			ep := &endpoint.Endpoint{
+				Targets: endpoint.Targets{
+					virtualServer.Status.VSAddress,
+				},
+				RecordType: "A",
+				DNSName:    virtualServer.Spec.Host,
+				Labels:     endpoint.NewLabels(),
+				RecordTTL:  ttl,
+			}
+
+			vs.setResourceLabel(virtualServer, ep)
+			endpoints = append(endpoints, ep)
+			continue
+		}
+
+	}
+
+	return endpoints, nil
+}
+
+// newUnstructuredConverter returns a new unstructuredConverter initialized
+func newVSUnstructuredConverter() (*unstructuredConverter, error) {
+	uc := &unstructuredConverter{
+		scheme: runtime.NewScheme(),
+	}
+
+	// Add the core types we need
+	uc.scheme.AddKnownTypes(f5VirtualServerGVR.GroupVersion(), &VirtualServer{}, &VirtualServerList{})
+	if err := scheme.AddToScheme(uc.scheme); err != nil {
+		return nil, err
+	}
+
+	return uc, nil
+}
+
+// filterByAnnotations filters a list of VirtualServers by a given annotation selector.
+func (vs *f5VirtualServerSource) filterByAnnotations(virtualServers []*VirtualServer) ([]*VirtualServer, error) {
+	labelSelector, err := metav1.ParseToLabelSelector(vs.annotationFilter)
+	if err != nil {
+		return nil, err
+	}
+
+	selector, err := metav1.LabelSelectorAsSelector(labelSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	// empty filter returns original list
+	if selector.Empty() {
+		return virtualServers, nil
+	}
+
+	filteredList := []*VirtualServer{}
+
+	for _, vs := range virtualServers {
+		// convert the VirtualServer's annotations to an equivalent label selector
+		annotations := labels.Set(vs.Annotations)
+
+		// include VirtualServer if its annotations match the selector
+		if selector.Matches(annotations) {
+			filteredList = append(filteredList, vs)
+		}
+	}
+
+	return filteredList, nil
+}
+
+func (vs *f5VirtualServerSource) setResourceLabel(virtualServer *VirtualServer, ep *endpoint.Endpoint) {
+	ep.Labels[endpoint.ResourceLabelKey] = fmt.Sprintf("f5-virtualserver/%s/%s", virtualServer.Namespace, virtualServer.Name)
+}
+
+//
+// The below structs and methods have been copy-pasted from
```

Comment:
The copy-pasting of these types and methods are unfortunate but tracked in https://github.com/F5Networks/k8s-bigip-ctlr/issues/2153, if there's a possibility to make the k8s API types importable.